035-rebalance-types-scaling-brokers.md

```diff
@@ -0,0 +1,118 @@
+
+In order to start a rebalancing on a cluster being scaled up or down, the proposal is about adding a `spec.mode` field in the `KafkaRebalance` custom resource with the following possible values:
+
+* `full-rebalance`: it starts a full-rebalance across all the brokers in the cluster as it works today. It is the default behavior if not specified (for backward compatibility).
+* `add-broker`: it starts a rebalancing by moving replicas out from the already existing brokers to the new added ones after a scale up.
+* `remove-broker`: it starts a rebalancing by moving replicas out from the target brokers to remove before a scale down.
+
+When the rebalancing is related to cluster being scaled up or down, the newly added brokers or the ones that are going to be removed have to be specified through a new `spec.brokers` field as an array.
+This needs come from how the Cruise Control API endpoints works. There is a specific query parameter which needs that list.
+
+A `KafkaRebalance` custom resource would look like this.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: <value> # full-rebalance, add-broker or remove-broker. Assuming full-rebalance by default if the field is not specified.
+  brokers:
+    - 3
+    - 4
+  goals:
+    - RackAwareGoal
+    - ReplicaCapacityGoal
+```
+
+### Flow
+
+The proposal doesn't assume any automation to run the rebalancing when the cluster is scaled up or down.
+Using the `StatefulSet` for Kafka brokers, the corresponding broker IDs are implied in one direction (added from the higher one up, removed from the higher one down) and the cluster operator could get them for the rebalancing.
+But on the other side, Strimzi is moving to use the [`StrimziPodSet`](https://github.com/strimzi/proposals/pull/44) and towards a ZooKeeper-less cluster support so that the broker IDs are handled differently and the user has the flexibility to scale down by removing specific brokers or scale up by specifying new broker IDs.
+For this reason the proposal is about having the user to specify the added or removed brokers for the rebalancing.
+If needed, it also provides the flexibility to scale up the cluster but running the rebalancing taking into account only a subset of the newly added brokers and leaving others just for new topics.
+The same could happen by running a rebalance for removing brokers to make them "empty" but then not doing the actual scale down.
+
+#### Rebalance on scale up
+
+When scaling up the cluster, the user can go through the following procedure:
+
+1. The user increases the number of replicas through the `spec.kafka.replicas` of the `Kafka` custom resource.
+2. When the scaling up is done, the user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `add-broker` and the list of the new brokers in the `spec.brokers` field.
```

Comment:
I don't think the user should have to do this. If CC is configured, then the user clearly wants its functionality. I think that Strimzi should create a KafkaRebalance CR with `spec.mode.add_broker` with all the fields filled in. Maybe even run the dry-run proposal (with the default CC settings) and have the results in the status, so that all the user has to do it approve it? The user could then, if they want, customize this auto-created KRCR with the other tuning options / goals and refresh it to get a new proposal.

This could be the standard behaviour and then we could have some flag in the Kafka cc config to turn that auto-KRCR creation functionality on/off.

I understand why we are still, at this stage, sticking with the user approval flow, but we should make things as easy as possible for users.